nxdrive/direct_edit.py

```diff
@@ -839,11 +844,28 @@ def _handle_upload_queue(self) -> None:
                 else:
                     # Try again in 30s
                     log.exception(f"Direct Edit unhandled HTTP error for ref {ref!r}")
-                    self._error_queue.push(ref)
+                    self._handle_upload_error(ref)
             except Exception:
                 # Try again in 30s
                 log.exception(f"Direct Edit unhandled error for ref {ref!r}")
-                self._error_queue.push(ref)
+                self._handle_upload_error(ref)
+
+    def _handle_upload_error(self, ref: Path) -> None:
+        key = str(ref)
+        print("handling error")
+        if key in self._upload_errors_watcher and self._upload_errors_watcher[key] == 3:
+            log.warning(f"Upload queue error: To many fails for {ref!r}")
+            self.directEditUploadFailed.emit(ref.name)
+            del self._upload_errors_watcher[key]
+            return
+        elif (
+            key in self._upload_errors_watcher and self._upload_errors_watcher[key] < 3
+        ):
+            self._upload_errors_watcher[key] += 1
+        else:
+            self._upload_errors_watcher[key] = 1
+        self._error_queue.push(ref)
+        print(self._upload_errors_watcher)
```

Comment:
Hm this code is not beautful to read. Maybe could we leverage [collections.defaultdict()](https://docs.python.org/3/library/collections.html#collections.defaultdict)?

Something like the following.

In `__init__()`, change to:
```python
    from collections import defaultdict  # move to the top, of course
    self._upload_errors_watcher: Dict[str, int] = defaultdict(int)
```

And adapt that method to something like:
```suggestion
    def _handle_upload_error(self, ref: Path) -> None:
        # (Add the missing docstring)
        self._upload_errors_watcher[ref] += 1
        if self._upload_errors_watcher[key] < 3:
            self._error_queue.push(ref)
            return

        log.error(f"Upload queue error: too many failures for {ref!r}, skipping the upload")
        self.directEditUploadFailed.emit(ref.name)  # TODO: adapt with directEditError
        self._upload_errors_watcher.pop(key, None)

```

Note: I did not test the code.